### ISSUE TITLE
New Structure for color themes

### ISSUE THREAD

**Issue by lansingthomas** [atlas-engineer/nyxt#3005]
@aartaka and I have some new surface colors in the oven, and I want to share the recipe here so you can say yes to it.
- feedback is appreciated, but not required.
- know that the next piece is revisiting the structure for typography (fonts per purpose), and this proposal will help with that.

### **Purpose**
1. establish a coherent structure for our default surface color theme, and future surface color themes; to help users see and interpret content, interact with the correct elements, and understand actions.
2. expand the color theme structure to cover all the current and future use cases we can think of. (so we don't need to hack in random values; to avoid future problems)
3. clarify the language around color themes.

### **Progress so far** 
Nyxt UI elements, potential new elements, css classes we might use later, and system states have been inventoried to assess [1] _how many colors do we require_ for a good UI, and [2] _how many color variations are required_ for each.
- the result of that inventory is that we require 8 colors and that each of those requires between 0 and 2 variations. 

We have been working to establish some programmatic norms around color selection to scale our choices today into future themes, and to aid with necessary typography selection.

### **What is your proposed change?**
The existing structure uses {background, primary, secondary, accent, warning} and an alternative for each. (10 colors total)

The better way is to structure the color themes by purpose, then provide 2 variations for each, for a total of 24 colors. like this:

#### 8 Main Colors
`primary`
- typically for: primary UI surfaces, panels etc.

`secondary`
- typically for: secondary UI elements, bars borders etc.

`background`
- typically for: large background elements

`warning`
- typically for: dangerous / potentially consequential actions, errors, failed downloads and page loads

`success`
- typically for: action completion, REPL successful run, successful/in-progress downloads, successful page loads

`action` (previously `accent`)
- typically for: focus affordance, action indicators, next-step prompting, background for selected elements

`codeblock` (previously `background-alt`)
- typically for: code, citation

`highlight` (as in #2994, possibly)
- typically for: highlighted text (selection), mission-critical information, search highlight, hints?

#### then we include the variations
 ONE tint, and ONE shade for each main color. This is to accomodate all the states of our widgets. For button it's:
- Regular state (main color.)
- Hovered state (shade color.)
- Pressed state (tint color.)
- Tab-focused state (shade surface or tint border.)

### **Rationale**
1. accommodate more complex UI components, we will require more colors for some elements (like dropdown menus with more states, and proper visual affordances for input fields), so let's add them now.
2. the current theme structure is somewhat arbitrary.
3. This proposal keeps the agreed on and brand ready colors from the old theme, and simply re-names them in helpful categories.
   - accent &rarr; `action`
   - background-alt &rarr; `codeblock`
   - primary, secondary, background, warning --- remain the same
4. YES contrast ratios have been considered, the logic from our surface structure here today also works with our text colors (`on-colors`) -- the on-color specifications will come with typography structure soon.
5. this will unblock that PR #2994 about text searching

### **Constraints**
- NO adding colors willy nilly. Artyom and I are meeting on Monday to try and use the new system to select color variations, and the proposed new colors `highlight` and `success`.
- NO changing brand guidelines; our purpose is to increase complexity and effectiveness without rethinking brand identity.
- NO breaking user created themes from previous versions.

# Pre-Build Checklist:
_talk through the change with someone, then come back and mark off what you can_
- [x] I have consulted with the team.
- [x] One or more developers have signed off on this change.
- [x] One or more user researchers have signed off on this change.

EDIT (by Artyom): Add more examples, fix some markup,

**Comment by aartaka** [comment]
> I'm really sorry to be bit negative here, but my impression is that we may be putting the horses in front of the case.
> 
> While the spec may look good on paper, I think we could benefit from letting the general case emerge from the particular ones.

While this is not a really general case, we have these small signals all over the place:
- The reason for the `background-alt` introduction in #2854 being new colors for code blocks in the manual.
- The introduction of "hightlight yellow" in #2994.
- Your fair statement about `diff-mode` having colors that're off-theme.

All of these lead me and Thomas to the idea that we need an exhaustive set of colors that covers all the semantics and system states there could be (and that's the reason why I don't really like the idea of the `codeblock` color—it seems quite weak semantically. But if we think about it as an "example" or "block" color, then it does make sense—we have lots of code blocks, quotes, or examples that we need to highlight in a specific way).

**Comment by aartaka** [comment]
> Let me remind everyone that we had a recent theme palette refactor and we are not even making full use of it (new `alt` and `warning` colors).

That's exactly why we actually don't need alternative colors—we need the sematic ones with established applications, instead of having a not-quite-the-same alternative to every color with uncertain application for it.

I see your point about the off-theme colors. But, much like @jmercouris appeal to authority (Apple), me and Thomas are relying on a closer, simpler, semantic, and much more battle-tested authority—Bootstrap. The reason the structure we're suggesting here is good is that it's mostly mimicks the Bootstrap one, and thus aims for universality.

**Comment by jmercouris** [comment]
I don't care why, just add the colors lmao :-D

**Comment by lansingthomas** [comment]
>I don't care why, just add the colors lmao :-D

yes! 

UPDATE:
Artyom and I spoke at length today, we chose a ton of magnificent colors, we will present them after we figure out the typography part. WIP scheduled for tomorrow.

**Comment by jmercouris** [comment]
**Proposed strategy:**

Add colors directly into the CSS *without* changing the theme palette. We are *NOT* going to mess with the theme library.

After we see a pattern emerge (before 4.0.0 for example), we consolidate the colors into a logical grouping and integrate them into the theme palette.

This way, we do not need to change any of the logic of the theme library.

**Current strategy:**

Involves anticipating all possible colors. Thinking about all logical ways that colors should be grouped.

**Reminder:**

We are all friends :-). Let's not get mad or upset!

**Why am I suggesting this?**

I do not want to discredit/invalidate the work that you are both doing! I like the fact that you are collaborating! I am just not certain it will yield the results you want. For this reason, I suggest we *timebox* the work you are doing! If we cannot integrate the new colors after one more meeting together, then probably we should adopt the newly proposed strategy so that we can make progress more quickly!

**Comment by jmercouris** [comment]
**Another option:**

Table the work you've done (to avoid it being wasted) and adopt the newly proposed strategy.

**Comment by lansingthomas** [comment]
I am happy to present... 
THE final, ultimate, never-again, forever, cut from stone, not to be reconsidered EVER Nyxt color theme structure!

Just kidding. But we do have what I hope will be an improvement. Including three `highlight` colors to work with for #2994, three `Action` colors to work with, and extensible typography guidelines.

Remember that **the purpose of this is to empower us as a team** (and our users!) to build better UI moving forward. We need a structure like this so that UI elements don't get all messy when users/contributors make new themes and pick new colors for themselves. The relationships between the colors are more important than the values, and this naming convention should help 

#### Surface colors (now with added colors!)
![surface colors june 2023 (1)](https://github.com/atlas-engineer/nyxt/assets/6666079/e7ea15df-ad4e-4b01-8fe2-3ee42cdb0e2d)


#### Typography Guidelines
![1440-900 light theme june 2023 (1)](https://github.com/atlas-engineer/nyxt/assets/6666079/5d772f2a-e502-4a04-a74b-c0654ea023af)



**Rationale for the  naming conventions and structure:**
TLDR; We offer tools, but not specifications or answers. 

Selecting and naming colors for a project like Nyxt involves tradeoffs. Some palettes have complete semantic descriptions of the colors for clarity in language and design, while others have fully contextual descriptions for each UI element (selected-button-border-on-hover...) to explicitly match color values with functionally specific elements across designs.   Some other systems even use both approaches in parallel.

_The scope of Nyxt does not require this_, and it risks constraining people in small but annoying ways as they implement their own themes and customizations.  
So!
We go with a semantically accurate color theme structure, that is both accessible and brand accurate, and we rely on our users to manage their own accessibility requirements If and when they stray outside of our themes. Hopefully the naming conventions will help people choose colors that will not cause messy interactions.

#### Big thank you to @aartaka for all the back and forth required to get this done.

**Comment by aartaka** [comment]
> Choose concepts that tell colors apart by contrast, not by how much light it contains.

Yes, that was exactly my point when prototyping the palette, but neither I nor Tom were able to come up with the theme-independent names. I'll be glad to call colors some name else, but I'm obviously restricted in what I can call them 🙃

> Sorry, but I'm afraid that the spec suffers from fundamental issues.

Whoa, slow down, names are in no way "fundamental"! If anything, names are the easiest thing to change about anything software. So let's come up with better names and stop throwing "fundamental" on something that merely lacks one easy adjustment 😉

**Comment by lansingthomas** [comment]
**How about this naming convention for surface colors?**
- _semantic.description_-less-contrast
- _semantic.description_
- _semantic.description_-more-contrast

**Would look like:**
![Surface Colors (light theme) (3)](https://github.com/atlas-engineer/nyxt/assets/6666079/c9a87ba5-4c56-4c4c-b7ff-374140355756)

**Rationale:**
the names communicate the relationships between the colors (so that users creating their own themes don't make too much of a mess)

**Comment by lansingthomas** [comment]
Uuugh, this is already breaking. Some surface colors can have both black and white text (looking at you Accent)

**Comment by lansingthomas** [comment]
the issue is that we are trying to make our UI kit behave as precise tokens for every case. I will make some tokens.

**Comment by lansingthomas** [comment]
Okay what If we do: `color primitives` + `token`s

**like this**
- the semantic information is in the `color primitive`
- the contextual information is in the `token`
   - the `token`s get value from the `color primitive`s
   - the `token`s get contextual information from `element state` {default, hovered, pressed, focused, disabled}
![color cards (1)](https://github.com/atlas-engineer/nyxt/assets/6666079/d50bb40d-bdce-461d-b08c-ecfd6d337af6)

**Comment by aartaka** [comment]
What if we actually called them `color-1`, `color-2`, `color-3` (with the contrast relations encoded in these, like "`color-1` is the least contrasting, `color-3` is the most contrasting of the three"). I mean, it's somewhat cheesy to resort to the numbered palettes of the other design systems (`*-50`-`*-900` from Bootstrap, Material Design, Dell Design System etc.), but it seems like the best linguistic device we can have.

Or even `color`, `color+`, and `color-`? Seems lispy enough, while conveying the idea. I mean, I myself kinda can imagine that `color+` is the more intensive/bright/contrasting variation of `color`, while `color-` is the less contrasting/eye-catching one.

And then, actual tokens basically use one of the numbered/plus/minus colors, as in
``` css
// Brightest on action
.button:active { background-color: theme:action+;}
// Less discernible hover
.button:hover { background-color: theme:action-;}
// Regular state
.button { background-color: theme:action;}
// Secondary disabled
.button:disabled { background-color: theme:secondary;}
```

-------------------------------------------------------

We also have our backs covered for the future in case we find this palette insuficient—we have `color++` to boast our colors in front of C programmers :D

**Comment by aartaka** [comment]
> Minor remark. The [spec](https://user-images.githubusercontent.com/6666079/250174114-e7ea15df-ad4e-4b01-8fe2-3ee42cdb0e2d.svg) above, features a `background` color that is not white. Why is that? A light theme needs to have a white color.

Not necessarily. A slight gray looks better than raw white (@lansingthomas probably knows the reasons behind that), much like slight saturated dark gray looks better than plain black for typography. And this slight deviation from white also allows us more flexibility for intense/dim/+/-/1/2/3 colors.

**Comment by aartaka** [comment]
> There are words that convey more contrast such as "intense"; and words that convey less contrast such as "dim" or "faint". But the naming doesn't matter much. The main requirement is that is must convey a concrete semantic idea for both light and dark themes so that when the user reads the snippet below he can visualize what will be the UI changes in both settings.

Yes. I like the intense/dim contrast too. I'm still too much into my +/- idea, but I can agree with this naming, especially compared to shade/tint.

> * The spec is only finished when we have palettes for both the light and dark theme versions;
> * For each of surface color, all variations of it must  preserve the same `on-color`.
> * Define a target contrast ratio internal for each variation of a surface color. Example: `action` must belong to [6.5, 7.5]; `action-more-contrast` must belong to [8.5, 9.5]; `actions-less-contrast` must belong to [4.5, 5.5].

I agree with all these. The exact contrast ratios we'll check when going through the palette, but it sounds like a reasonable set of defaults.

**Comment by lansingthomas** [comment]
> Minor remark. The [spec](https://user-images.githubusercontent.com/6666079/250174114-e7ea15df-ad4e-4b01-8fe2-3ee42cdb0e2d.svg) above, features a background color that is not white. Why is that? A light theme needs to have a white color.
- off-white background colors and off-black text colors are really helpful for visual communication as we layer panels and differentiate states. Sometimes the difference is barely perceptible, but still important.
- off-white accommodates the active pseudo-class for already bright elements
- Don't worry #FFFFFF is still on our workbench

**Comment by lansingthomas** [comment]
Okay here are the colors for light theme.
- with a simple name that tells us about contrast direction using +/-

![Surface Colors (light theme) (4)](https://github.com/atlas-engineer/nyxt/assets/6666079/6d567c2f-68db-49f5-9572-9bcc8fc1604b)

Please check box, or provide new names
- [ ] artyom agrees with color names
- [ ] andre agrees with color names

---

I will elaborate interaction states for the light theme.  Which will look something like this:
<img width="605" alt="Screenshot 2023-07-04 at 2 57 1" src="https://github.com/atlas-engineer/nyxt/assets/6666079/9b7520c0-6fc0-448e-8ff7-151c8f26a46b">

Then @aartaka and I will get the dark theme up to speed (and come up with a plan for tokens)

**Comment by lansingthomas** [comment]
> there may be some mistakes in your palette. Example: the on-color corresponding to secondary in the light theme is black, not white. You can tell immediately by how hard it is to read.

Oops. good catch.

### text colors
![Text Colors - Light Theme](https://github.com/atlas-engineer/nyxt/assets/6666079/0a24f8cc-b4c4-4b0e-947b-b00c95448ea7)

---

### Typography Guide: Light Theme
![Typography Guide](https://github.com/atlas-engineer/nyxt/assets/6666079/5d189ae7-eea3-4e07-a8fe-1f20af5ee9aa)

---
### Surface Colors: Light Theme

![Surface Colors (light theme) (5)](https://github.com/atlas-engineer/nyxt/assets/6666079/909fbbf7-a42e-40f7-9b88-385bb4fe337b)

**Comment by lansingthomas** [comment]
Do these color names seem sensible? 
**bright-text** | for white 
**dark-text1** | typically for black text in _subdued_ interaction states
**dark-text** | for most black text (:default)
**dark-text3** | typically for black text in _intense_ interaction states

I know we have been using the  _on-color_ convention.

the proposed names for text colors use 1,2,3 instead of +/- because the contrast that +/- communicates is relative to the text colors. (I don't want to give the impression that text colors can be treated the same as surface colors)

**Comment by lansingthomas** [comment]
I agree that constraints are helpful at this stage, and that less complexity is the way to move forward...

However, We do still require dark text variations for good visual communication in our documentation. I promise it will be helpful. Top blogs on Hacker News do it, legacy media like NYT and Al Jazeera do it. 

The values may change but I need 3 black text variables.

**Comment by lansingthomas** [comment]
The main thing is having header text and paragraph text as different shades &rarr; skimmability in large paginating docs like our documentation

**Comment by lansingthomas** [comment]
added px (density independent) and corrected a few small errors.
![Typography Guide (1)](https://github.com/atlas-engineer/nyxt/assets/6666079/3f76444d-1076-4813-98c5-be29f3097b49)

**Comment by jmercouris** [comment]
Well, we don't use title case simply because it is ambiguous. I actually prefer title case aesthetically, but I was won over by this logic. In this way we don't have to think about what to capitalize and what not to capitalize.

With regards to shipping Nyxt with fonts, actually this is OK, we have talked about, just technically hard to do. To ship them and load them locally.

**Comment by lansingthomas** [comment]
I am still on the fence myself about _Sentence case_ and _Title Case_

This thread is getting sort of long. I think we need to split it up to start locking pieces in place.

I will start a new issue for Typography Guidelines.

**Comment by aartaka** [comment]
> 1 I believe that Nyxt has decided to never follow title case. Personally, I like title case but the team seems to have settled on not using it for a long
>  time. I.e. they prefer "A bag of potatoes" over "A Bag of
>  Potatoes". @jmercouris can confirm it.

As @jmercouris already said, that's more of a historic accident than the rule.

Title case is too customary in English prose to ignore it. Seeing a
heading in sentence case kinda looks strange after some time with title
case blogosphere for me.

**Comment by jmercouris** [comment]
Well, it wasn't exactly an accident, it is Pierre's preference :-P

**Comment by lansingthomas** [comment]
Im leaning toward Title Case for H1 and Sentence case for H2-4

# Because It Looks
## Pretty okay
### Don't you think?

**Comment by lansingthomas** [comment]
UPDATE: Its gonna be beautiful
- @aartaka and I got together today and mashed all the junk from this thread with all the junk in our brains.
   - &rarr; light theme seems good to go
   -  &rarr; found a way to preserve the logical goodness of on-colors, while also giving us enough text-colors to work with for typography and pseudo-classes for interact-able elements.
   - &rarr; dark theme is really close but there are some contrast issues in one category that need resolved before publication. (highlight, highlight-, and highlight+)

---

## Text colors for both themes
1. the text-colors &rarr; on-colors
2. on-colors are paired with surface colors
3. the extra tints/shades are for long docs with hierarchy

![Text Colors - Dark Theme (Amphora)](https://github.com/atlas-engineer/nyxt/assets/6666079/2436985b-714c-4e43-a2b6-0ee6848eff4a)
![Text Colors - Light Theme (2)](https://github.com/atlas-engineer/nyxt/assets/6666079/28a36881-9316-471c-b055-b443a16ee054)

## The themes
![Surface Colors (Dark Theme)](https://github.com/atlas-engineer/nyxt/assets/6666079/7ad33415-b008-41b4-8cc9-affc88ac6d2e)
![Surface Colors (light theme) (7)](https://github.com/atlas-engineer/nyxt/assets/6666079/92b2fa31-c725-4e58-85d9-732139e4f501)

**Comment by jmercouris** [comment]
I really like how these colors have come out, I'm eager to see them in the actual interface! :-)

**Comment by lansingthomas** [comment]
Huray! the structure seems to work well. 

Here are the corrected themes (in one place) with some prompt buffer mockups.
- now featuring contrast ratios within specification :D
  - the spec:  _{greater than 4.5, for reduced contrast elements, greater than 6.5 for standard elements, greater than 8.5 for increased contrast elements}_

# Light theme surface colors
![Surface Colors (light theme) (8)](https://github.com/atlas-engineer/nyxt/assets/6666079/5e2a24d8-256a-4c4e-9ffc-e44f604bfba3)

### would look like:
![Prompt buffer light theme color testing](https://github.com/atlas-engineer/nyxt/assets/6666079/4fcf617d-1227-4a37-a522-7781aea8b508)

# Dark Theme surface colors
![Surface Colors (Dark Theme) (1)](https://github.com/atlas-engineer/nyxt/assets/6666079/ca5384a0-324a-4039-911c-ee33c7af7c99)

### would look like:
![Prompt buffer dark theme color testing](https://github.com/atlas-engineer/nyxt/assets/6666079/04fb310d-d5af-4e46-b6c3-5f95ad339c19)

# the text-colors
same as above.
@aartaka has a plan for implementing this part. 

# Interaction State Guidelines for other elements
![Interation State Guidelines](https://github.com/atlas-engineer/nyxt/assets/6666079/e1490348-2eb1-4ed5-a207-d3362ab84706)

Thank you @aartaka for your help!

**Comment by lansingthomas** [comment]
P.S. the interaction state guidelines are only for light theme at this stage.

**Comment by aartaka** [comment]
I had an idea for technical design to preserve both current API and the new one working so that we don't break too much things. But I'll wait for the PR to arrive, because @aadcg was first to assign 😝

**Comment by aartaka** [comment]
OK, will describe the implementation on Monday.

**Comment by aartaka** [comment]
@lansingthomas, we forgot to discuss the semantics of fonts (heading font, text font etc.)! I just realized that we don't have a plan for our theme `font-family`, which is used exclusively for headings at the moment.

**Comment by aartaka** [comment]
I've started theme restructuring for surface and on-* colors, preserving the backwards compatibility whenever possible (=everywhere). Will open a PR in a couple of days. Happens on a `theme-restructure` branch in case y'all are curious.

**Comment by lansingthomas** [comment]
> we forgot to discuss the semantics of fonts (heading font, text font etc.)! I just realized that we don't have a plan for our theme font-family, which is used exclusively for headings at the moment.

Yes!
Lets do a separate issue for that. I have been thinking about it. Lets find a time to screen share tomorrow.

**Comment by lansingthomas** [comment]
Dark theme (edited to fix a mistake with one of the warning colors)
![Surface Colors (Dark Theme) (2)](https://github.com/atlas-engineer/nyxt/assets/6666079/b2b29cee-6c49-485f-a677-ac14285a0d2d)
